docs/syntax.md

```diff
@@ -0,0 +1,667 @@
+
+### Tuple
+
+```python
+(3, 9)
+```
+
+A Tuple is surrounded by `(` `)` and should contain 2 or more elements. Elements are separated by `,`.
+
+### Superposition
+
+```python
+{1 2 3}
+```
+
+A superposition of values is defined using `{` `}` with at least 2 expressions inside.
+
+### Numbers and Infix Operations
+
+Currently, bend supports 3 types of numbers: floats, integers and unsigned integers. All of then are 24 bit sized.
+
+```python
+f24 = +88.012
+
+i24 = -42
+
+u24 = 42
+```
+
+| Operation      | Syntax |
+|----------------|--------|
+| Addition       | x + y  |
+| Subtraction    | x - y  |
+| Multiplication | x * y  |
+| Division       | x / y  |
+| Equal          | x == y |
+| Not Equal      | x != y |
+| Less Than      | x < y  |
+| Greater Than   | x > y  |
+| Bitwise And    | x & y  |
+| Bitwise Or     | x | y  |
+| Bitwise Xor    | x ^ y  |
+
+### Character Literal
+
+```python
+'x'
+```
+
+A Character is surrounded with `'`. Accepts unicode characters, unicode escapes in the form '\u{hex value}' and is desugared to the unicode codepoint as an `u24`.
+
+Only supports unicode codepoints up to `0xFFFFFF`.
+
+### String Literal
+
+```python
+"Hello, World!"
+```
+
+A String literal is surrounded with `"`. Accepts the same values as characters literals.
+
+It is desugared to constructor calls of the built-in type String, `String/cons(head, ~tail)` and `String/nil` .
+
+### List Literal
+
+```python
+[1, 2, "three"]
+```
+
+A List literal is surrounded by `[` `]`. The elements must be separated by `,`.
+
+It is desugared to constructor calls of the built-in type List, `List/cons(head, ~tail)` and `List/nil` .
+
+### List Comprehension
+
+```python
+[x + 1 for x in list]
+
+[x + 1 for x in list if x > 2]
+```
+
+A List Comprehension generates a new list, it can be extracted in 3 parts.
+
+`[expression . iterator . condition]`
+
+Expression: The expression to be performed in the iterator element.
+
+Iterator: Binds a name to the list elements.
+
+Condition: Optional, is used to filter the list elements.
+
+It is desugared to a fold statement:
+
+```python
+fold list:
+  List/cons:
+    if condition:
+      List/cons(list.head, list.tail)
+    else:
+      list.tail
+  List/nil:
+    List/nil
+```
+
+<div id="core-syntax"></div>
+
+# Core Syntax
+
+## Top-level definitions
+
+```rust
+data Name
+  = (Ctr1 arg1 arg2)
+  | Ctr2
+
+Name (Ctr1 sub_arg1 sub_arg2) arg3 = rule0_body
+Name Ctr2 arg3 = rule1_body
+```
+
+### Definitions
+
+Definitions can have multiple rules pattern matching each defined argument.
+
+```rust
+identity x = x
+
+Bool.neg True  = False
+Bool.neg False = True
+
+MapMaybe (Some val) f = (Some (f val))
+MapMaybe None f = None
+
+Pair.get (fst, snd) f = (f fst snd)
+```
+
+A function definition is composed of a sequence of rules, where a rule is the name of the function, a sequence of patterns and then the body.
+
+A rule pattern can be:
+- A variable.
+- A number.
+- A constructor.
+- A tuple.
+- A superposition.
+
```

Comment:
Add a link to the document on pattern matching to learn about what exactly the rules for pattern matching equations are